gap/congruences/congsemigraph.gi

```diff
@@ -0,0 +1,279 @@
+        for e in gens do
+          if Source(e) = verts[v] and Range(e) = verts[u] then
+            Add(pairs, [verts[v], e * e ^ -1]);
+          fi;
+        od;
+      fi;
+    od;
+  od;
+  return SemigroupCongruence(S, pairs);
+end);
+
+BindGlobal("SEMIGROUPS_MinimalHereditarySubsetsVertex",
+function(D, v)
+  local subsets, hereditary, u, out, j, s, a;
+  if not (v in DigraphVertices(D)) then
+    ErrorNoReturn(v, " must be a vertex of ", D);
+  fi;
+  out := Set(OutNeighboursMutableCopy(D)[v]);
+  subsets := [];
+  for u in [1 .. Length(out)] do
+    a := out[u];
+    RemoveSet(out, a);
+    hereditary := Set([]);
+    Append(hereditary, out);
+    for s in out do
+      for j in VerticesReachableFrom(D, s) do
+        AddSet(hereditary, j);
+      od;
+    od;
+    if not (hereditary in subsets) and not (a in hereditary) then
+      Add(subsets, hereditary);
+    fi;
+    AddSet(out, a);
+  od;
+  return AsSortedList(subsets);
+end);
+
+InstallMethod(GISGeneratingCongruences,
+"for a graph inverse semigroup",
+[IsGraphInverseSemigroup],
+function(S)
+  local congs, out, v, h, D;
+  D := GraphOfGraphInverseSemigroup(S);
+  congs := [];
+  out := OutNeighbours(D);
+  for v in DigraphVertices(D) do
+    if Length(out[v]) = 0 then
+      Add(congs, CongruenceByWangPair(S, [v], []));
+    elif Length(out[v]) = 1 then
+      Add(congs, CongruenceByWangPair(S, [], [v]));
+    else
+      for h in SEMIGROUPS_MinimalHereditarySubsetsVertex(D, v) do
+        Add(congs, CongruenceByWangPair(S, h, [v]));
+      od;
+    fi;
+  od;
+  Add(congs, CongruenceByWangPair(S, [], []));
+  return congs;
+end);
+
+InstallMethod(AsCongruenceByWangPair, "for a semigroup congruence",
+[IsSemigroupCongruence],
+function(C)
+  local H, W, eq, j, Hgraph, Wgraph, v;
+  Hgraph := [];
+  Wgraph := [];
+  if not IsGraphInverseSemigroup(Source(C)) then
+    ErrorNoReturn(Source(C), " is not a graph inverse semigroup");
+  fi;
+
+  H := [];
+  W := [];
+  eq := EquivalenceRelationPartition(C);
+  eq := Filtered(eq, x -> ForAny(x, IsVertex));
+  for j in eq do
+    if MultiplicativeZero(Source(C)) in j then
+      Append(H, Filtered(j, IsVertex));
+    else
+      Append(W, Filtered(j, IsVertex));
+    fi;
+  od;
+  Sort(H);
+  Sort(W);
+  for v in H do
+    Add(Hgraph, IndexOfVertexOfGraphInverseSemigroup(v));
+  od;
+  for v in W do
+    Add(Wgraph, IndexOfVertexOfGraphInverseSemigroup(v));
+  od;
+  return CongruenceByWangPair(Source(C), Hgraph, Wgraph);
+end);
+
+InstallMethod(JoinSemigroupCongruences,
+"for two congruences by Wang pair",
+[IsCongruenceByWangPair, IsCongruenceByWangPair],
+function(cong1, cong2)
+  local out, H, W, v, c, u, X, W_zero, S, D, w, e, k, d, H1, H2;
+  S := Source(cong1);
+  D := GraphOfGraphInverseSemigroup(S);
+  out := OutNeighbours(D);
+  X := [];
+  H1 := cong1!.H;
+  H2 := cong2!.H;
+  H := Union(H1, H2);
+  W := Difference(Union(cong1!.W, cong2!.W), H);
+  W_zero := [];
+  for v in W do
+    c := 0;
+    for w in out[v] do
```

Comment:
if not ForAll(out[v], w -> w in H) then

Reply:
`if ForAll(out[v], w -> w in H) then` :)